Enable local-up-cluster.sh to start with the secured port

dims opened the pull request:
related to #33375

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33966)

<!-- Reviewable:end -->

dims committed:
Enable local-up-cluster.sh to start with the secured port

Generate a kubeconfig for use with controller-manager, kubelet,
scheduler etc. This kubeconfig should use the secure https
port of the api server with appropriate ca cert for the components
to talk to api server.

With this change, one can set API_PORT=0 to completely switch off
insecure access for testing admission controllers etc.

Fixes #33375

eparis commented on hack/local-up-cluster.sh:
```diff
@@ -139,15 +140,18 @@ CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
     # For the common local scenario, fail fast if server is already running.
     # this can happen if you run local-up-cluster.sh twice and kill etcd in between.
-    curl --silent -g $API_HOST:$API_PORT
+    curl --silent -k -g $API_HOST:$API_SECURE_PORT
```
Why not use `--cacert=$(ROOT_CA_FILE)` instead of `-k` ?

eparis commented on hack/local-up-cluster.sh:
```diff
@@ -545,8 +568,9 @@ To start using your cluster, open up another terminal/tab and run:
 
-  cluster/kubectl.sh config set-cluster local --server=http://${API_HOST}:${API_PORT} --insecure-skip-tls-verify=true
-  cluster/kubectl.sh config set-context local --cluster=local
+  cluster/kubectl.sh config set-cluster local --server=https://${API_HOST}:${API_SECURE_PORT} --insecure-skip-tls-verify=true
```
do we still need the ``insecure-skip-tls-verify` ?

dims replied:
yes, cleaning up now

dims commented:
@k8s-bot kubemark e2e test this

dims commented:
@k8s-bot gci gce e2e test this

dims commented:
ok, here's what i got working so far:

1. Run hack/local-up-cluster.sh, then stop it with CTRL-C
2. disable the insecure port using `export API_PORT=0`
3. Run hack/local-up-cluster.sh

Essentially apiserver.key/apiserver.crt are generated in step #1 and the local client can use the apiserver.crt. 

Note that we create the local client first and then generate the certs when we start the secure listener. Hence the two step tango. Any ideas to make this better? Please let me know!

dims commented:
@deads2k : your thoughts please? (See my last comment)

deads2k commented:
I think that this will order the parameters in a way that allows it to work on the first try: https://github.com/kubernetes/kubernetes/pull/34556 .

dims commented:
We need #34556 to merge first

deads2k commented:
@dims I definitely want this, but I'm not sure it completely fixes #33375 since the `kube-controller-manager` won't have an identity, right?

This pull should still merge before that (it's a big step forward), but I don't think we're all the way there yet.

dims commented:
@deads2k yep, indeed. we can leave the #33375 open and leave notes as to what more steps are needed.

deads2k commented on hack/local-up-cluster.sh:
```diff
@@ -139,15 +140,18 @@ CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
+    curl --silent -k -g $API_HOST:$API_SECURE_PORT
     if [ ! $? -eq 0 ]; then
-        echo "API SERVER port is free, proceeding..."
+        echo "API SERVER secure port is free, proceeding..."
```
probably ought to check both.

dims replied:
Ack. Fixing

dims commented on pkg/genericapiserver/options/server_run_options.go:
```diff
@@ -227,8 +228,21 @@ func (s *ServerRunOptions) NewSelfClientConfig(token string) (*restclient.Config
-	if s.SecurePort > 0 && len(s.TLSCAFile) > 0 {
-		clientConfig.Host = "https://" + net.JoinHostPort(s.BindAddress.String(), strconv.Itoa(s.SecurePort))
+
+	// Use secure port if the TLSCAFile is specified and the file exists
```
Right, removing the fallback to insecure

dims commented:
@eparis @deads2k Addressed all the comments and tested `API_PORT=0` locally after cleaning up all artifacts (`sudo rm -rf /var/run/kubernetes/ /tmp/kube* ~/.kube/`)

deads2k commented:
`API_PORT=0 hack/local-up-cluster.sh` is ending in `!!! [1014 10:36:12] Timed out waiting for apiserver:  to answer at https://localhost:6443/api/v1/pods; tried 10 waiting 1 between each
Cleaning up...
` for me

dims commented:
@deads2k can you paste me your /tmp/kube-apiserver.log please?

deads2k commented:
> @deads2k can you paste me your /tmp/kube-apiserver.log please?

https://gist.github.com/deads2k/6181c16615618daeb7b3d6fdb8e16f72

dims commented:
@deads2k @sttts - found a crumb to solve the mystery. seems curl behaves differently with the certs we generate when using `--cert-dir`

```
[fedora@my-fedora kubernetes]$ curl -v --cacert /var/run/kubernetes/apiserver.crt -gfs https://localhost:6443/api/v1/pods
*   Trying ::1...
* Connected to localhost (::1) port 6443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /var/run/kubernetes/apiserver.crt
  CApath: none
* Server certificate:
* 	subject: CN=149.202.163.230@1476827403
* 	start date: Oct 18 21:50:03 2016 GMT
* 	expire date: Oct 18 21:50:03 2017 GMT
* 	common name: 149.202.163.230@1476827403
* 	issuer: CN=149.202.163.230@1476827403
* NSS error -8101 (SEC_ERROR_INADEQUATE_CERT_TYPE)
* Certificate type not approved for application.
* Closing connection 0
```

dims commented:
So for now i'll have to switch `hack/lib/util.sh` to just use `-k` instead of `--ca-cert`

sttts commented:
@dims the cert error with curl comes from the "X509v3 Extended Key Usage" field in the cert. Certs with this attribute (which includes the usage as server cert in our case) cannot be used as CA with curl (and Firefox afaik). The solution is to create a root CA with one key and another key for the actual apiserver cert. This would be much cleaner IMO.

dims commented:
@sttts are we talking about fixing --cert-dir automatic generation of certs (apiserver.key/apiserver.crt). Adding an additional root CA file there?

Or are we talking about using say cfssl to generate stuff so we can pass in --tls-* flags (instead of using --cert-dir)?

Either way, can we do that in next/another PR?

Thanks,
Dims

sttts commented:
@dims my guess is that the auto generated cert is broken for curl under certain circumstances. Of course, using cfssl would be the way to go for a rootca+apiserver.crt implementation, but this can wait for another PR.

I tested your changes on my machine with success. Still wondering why it doesn't work for @deads2k. Am on Fedora 24 here.

I saw that curl error before when I had a broken API_HOST=0.0.0.0 in my environment. While troubleshooting I was able to make curl happy with https://gist.github.com/sttts/b45105abe7779f89aa75ef90638a48c7.

deads2k commented:
If it works for @sttts I'll merge as is and hack on it later.

liggitt commented:
the auto-generated cert is a CA *and* a server cert

dims commented:
@liggitt agree. it's just that it does not work well on fedora. One option seems to be that we should do as @sttts suggests (https://gist.github.com/sttts/b45105abe7779f89aa75ef90638a48c7)

sttts commented:
@liggitt @dims did a rough cfssl implementation with root-ca today: https://github.com/kubernetes/kubernetes/pull/35143. Curl and go's tls stack are happy.

dims commented:
@sttts very cool. that was fast 👍
